app/src/androidTest/java/com/example/sharingang/UserProfileFragmentTest.kt

```diff
@@ -21,11 +26,51 @@ class UserProfileFragmentTest {
     @get:Rule(order = 1)
     val activityRule = ActivityScenarioRule(MainActivity::class.java)
 
+    @get:Rule(order = 2)
+    val permissionRule: GrantPermissionRule = GrantPermissionRule.grant(
+        Manifest.permission.READ_EXTERNAL_STORAGE,
+        Manifest.permission.CAMERA
+    )
+
     @Test
     fun canOpenSearchFragment() {
         navigate_to(R.id.userProfileFragment)
 
         val textView = onView(withId(R.id.nameText))
         textView.check(matches(withText(FakeCurrentUserProvider.fakeUser.name)))
     }
+
+    @Test
+    fun pictureButtonsAreDisplayedCorrectly() {
+        navigate_to(R.id.userProfileFragment)
+        onView(withId(R.id.btn_open_camera)).check(matches(withText("Open Camera")))
+        onView(withId(R.id.btn_open_gallery)).check(matches(withText("Open Gallery")))
+        onView(withId(R.id.btnApply)).check(matches(not(isDisplayed())))
+    }
+
+
+
+    @Test
+    fun applyButtonIsDisplayedUponClickOnOpenGallery() {
+        val device: UiDevice = UiDevice.getInstance(getInstrumentation())
+        navigate_to(R.id.userProfileFragment)
+        onView(withId(R.id.btn_open_gallery)).perform(click())
+        device.pressBack()
+        onView(withId(R.id.btnApply)).check(matches(isDisplayed()))
+        onView(withId(R.id.btnApply)).perform(click())
+        onView(withId(R.id.btnApply)).check(matches(not(isDisplayed())))
+    }
+
+    @Test
+    fun applyButtonApplies() {
+        val device: UiDevice = UiDevice.getInstance(getInstrumentation())
+        navigate_to(R.id.userProfileFragment)
+        onView(withId(R.id.btn_open_gallery)).perform(click())
+        device.pressBack()
+        onView(withId(R.id.btnApply)).check(matches(isDisplayed()))
+        UserProfileFragment.isTestCase = true
+        onView(withId(R.id.btnApply)).perform(click())
+        onView(withId(R.id.btnApply)).check(matches(not(isDisplayed())))
+        UserProfileFragment.isTestCase = false
```

Comment:
Isn't there a way to test the class without using attributes made specifically for testing purposes?

Reply:
I can remove it but for now I have not found any other way to test the gallery or the camera's response, even with Intent stubbing there are too many unknowns because we are using a more or less custom interface for image support. I guess the options would be either to keep it as it is, or remove it but the coverage would go down.